From 6c8c055d247ef011d70e2220b6273056b3dcdd64 Mon Sep 17 00:00:00 2001
From: Yong Yao <yong.yao@intel.com>
Date: Sat, 15 Jun 2013 14:15:34 +0800
Subject: opengl: Implement fast tls forwarders for x86

EGL doesn't support C only extension forwarders easily. To enabled all
EGL extensions from underlying driver we have to implement assembler
functions that forward calls to underlying HW driver functions or to
android functions.

Functions have to have noinline attribute to prevent label conflict in
the assembler code.

Inline assembler has to rollback function prelogue before calling
underlying function because x86 gcc doesn't support naked functions.

Change-Id: I723075eb43a8966f6244f27624d7aa48b1d3e531
Original-Change-Id: I4b046281f7307dd647731a48fa92319fdbab3090
Signed-off-by: Pauli Nieminen <pauli.nieminen@linux.intel.com>
Signed-off-by: Sean V Kelley <sean.v.kelley@linux.intel.com>
Signed-off-by: Yong Yao <yong.yao@intel.com>
---
 opengl/libs/EGL/getProcAddress.cpp |   24 ++++++++++++++++++++++++
 opengl/libs/GLES2/gl2.cpp          |   21 +++++++++++++++++++++
 opengl/libs/GLES_CM/gl.cpp         |   21 +++++++++++++++++++++
 opengl/libs/hooks.h                |    2 +-
 4 files changed, 67 insertions(+), 1 deletions(-)

diff --git a/opengl/libs/EGL/getProcAddress.cpp b/opengl/libs/EGL/getProcAddress.cpp
index c160aa0..8a7b8ac 100644
--- a/opengl/libs/EGL/getProcAddress.cpp
+++ b/opengl/libs/EGL/getProcAddress.cpp
@@ -89,6 +89,30 @@ namespace android {
                 :                                                   \
             );
 
+    #elif defined(__i386__)
+
+        #define API_ENTRY(_api) __attribute__((noinline))  _api
+
+        /* pop %%ebp only because x86 doesn't support naked functions */
+        #define CALL_GL_EXTENSION_API(_api)                         \
+            asm volatile(                                           \
+                "pop %%ebp                  \n"                     \
+                "movl %%gs:0, %%ecx         \n"                     \
+                "movl %c[tls](%%ecx), %%eax \n"                     \
+                "test %%eax, %%eax          \n"                     \
+                "je out" #_api "            \n"                     \
+                "movl %c[api](%%eax), %%ecx \n"                     \
+                "test %%ecx, %%ecx          \n"                     \
+                "je out" #_api "            \n"                     \
+                "jmp *%%ecx                 \n"                     \
+                "out" #_api ": ret          \n"                     \
+                :                                                   \
+                : [tls] "i"(TLS_SLOT_OPENGL_API*4),                 \
+                  [api] "i"(__builtin_offsetof(gl_hooks_t,          \
+                                          ext.extensions[_api]))    \
+                :                                                   \
+                );
+
     #else
         #error Unsupported architecture
     #endif
diff --git a/opengl/libs/GLES2/gl2.cpp b/opengl/libs/GLES2/gl2.cpp
index fad2176..3eb1ab0 100644
--- a/opengl/libs/GLES2/gl2.cpp
+++ b/opengl/libs/GLES2/gl2.cpp
@@ -94,6 +94,27 @@ using namespace android;
             :                                                    \
             );
 
+    #elif defined(__i386__)
+
+        #define API_ENTRY(_api) __attribute__((noinline)) _api
+
+        /* pop %%ebp only because x86 doesn't support naked functions */
+        #define CALL_GL_API(_api, ...)                              \
+            asm volatile(                                           \
+                "pop %%ebp                  \n"                     \
+                "movl %%gs:0, %%ecx         \n"                     \
+                "movl %c[tls](%%ecx), %%eax \n"                     \
+                "test %%eax, %%eax          \n"                     \
+                "je out" #_api "            \n"                     \
+                "movl %c[api](%%eax), %%ecx \n"                     \
+                "jmp *%%ecx                 \n"                     \
+                "out" #_api ": ret          \n"                     \
+                :                                                   \
+                : [tls] "i"(TLS_SLOT_OPENGL_API*4),                 \
+                  [api] "i"(__builtin_offsetof(gl_hooks_t, gl._api))    \
+                :                                                   \
+                );
+
   #else
 
     #error Unsupported architecture
diff --git a/opengl/libs/GLES_CM/gl.cpp b/opengl/libs/GLES_CM/gl.cpp
index a5bbdc6..4eac9a4 100644
--- a/opengl/libs/GLES_CM/gl.cpp
+++ b/opengl/libs/GLES_CM/gl.cpp
@@ -149,6 +149,27 @@ GL_API void GL_APIENTRY glWeightPointerOESBounds(GLint size, GLenum type,
             :                                                    \
             );
 
+    #elif defined(__i386__)
+
+        #define API_ENTRY(_api) __attribute__((noinline)) _api
+
+        /* pop %%ebp only because x86 doesn't support naked functions */
+        #define CALL_GL_API(_api, ...)                              \
+            asm volatile(                                           \
+                "pop %%ebp                  \n"                     \
+                "movl %%gs:0, %%ecx         \n"                     \
+                "movl %c[tls](%%ecx), %%eax \n"                     \
+                "test %%eax, %%eax          \n"                     \
+                "je out" #_api "            \n"                     \
+                "movl %c[api](%%eax), %%ecx \n"                     \
+                "jmp *%%ecx                 \n"                     \
+                "out" #_api ": ret          \n"                     \
+                :                                                   \
+                : [tls] "i"(TLS_SLOT_OPENGL_API*4),                 \
+                  [api] "i"(__builtin_offsetof(gl_hooks_t, gl._api))    \
+                :                                                   \
+                );
+
   #else
     #error Unsupported architecture
   #endif
diff --git a/opengl/libs/hooks.h b/opengl/libs/hooks.h
index b2a684c..7d59606 100644
--- a/opengl/libs/hooks.h
+++ b/opengl/libs/hooks.h
@@ -32,7 +32,7 @@
 #include <GLES3/gl3.h>
 #include <GLES3/gl3ext.h>
 
-#if !defined(__arm__) && !defined(__mips__)
+#if !defined(__arm__) && !defined(__mips__) && !defined(__i386__)
 #define USE_SLOW_BINDING            1
 #else
 #define USE_SLOW_BINDING            0
-- 
1.7.4.1

